### PR TITLE
[krb5] initial integration

### PR DESCRIPTION
**Pull request by 0x34d**
Hello @krb5 and Team @tlyu @greghudson @kaduk
Can you comment on ```krb5``` integration into ```oss-fuzz``` for ```security and fuzzing``` of krb5?
I and the ```oss-fuzz``` team would like to hear your thought on integration.

Signed-off-by: 0x34d <ajsinghyadav00@gmail.com>

### CODE REVIEW

**Comment by greghudson** [comment]
Things I noted:

* --enable-static --disable-shared is a barely supported pair of configure options, which disables large swaths of functionality.  Are they needed?

* What are the criteria for inclusion in vendor_ccs in project.yaml?  Putting individual addresses there seems wrong, and of the two people currently listed, only I am currently paid by MIT to work on krb5.

* The surfaces chosen seem productive insofar as they won't just repeatedly hammer at authenticated decryption (as something like krb5_rd_req() would).  But they also don't cover very much of the exposed krb5 protocol interface, and I can't think of any recent CVEs associated with those surfaces.

**Comment by tlyu** [comment]
Hi! I no longer work on @krb5 for MIT, for over 5 years now. kaduk has been gone for even longer. Please explain how you decided which individuals to contact or list. Thanks.

**Comment by 0x34d** [comment]
> Hi! I no longer work on @krb5 for MIT, for over 5 years now. kaduk has been gone for even longer. Please explain how you decided which individuals to contact or list. Thanks.

Thanks for the update. @tlyu

**Comment by 0x34d** [comment]
>

**Comment by 0x34d** [comment]
I'm Stupid i click the close button. 
abusive words, abusive words, abusive words.

**Comment by 0x34d** [comment]
> * --enable-static --disable-shared is a barely supported pair of configure options, which disables large swaths of functionality.  Are they needed?

Yes, For the fuzzer to monitor the Application, it needed to be compiled with the static library.

> * What are the criteria for inclusion in vendor_ccs in project.yaml?  Putting individual addresses there seems wrong, and of the two people currently listed, only I am currently paid by MIT to work on krb5.

Anyone who is working on the Application(krb5). and take care of bug reports from oss-fuzz.

> * The surfaces chosen seem productive insofar as they won't just repeatedly hammer at authenticated decryption (as something like krb5_rd_req() would).  But they also don't cover very much of the exposed krb5 protocol interface, and I can't think of any recent CVEs associated with those surfaces.

Thanks for guiding me for ```krb5_rd_req()``` .
I will create a harness as soon as possible.

**Review by Alan32Liu — APPROVED** [review]
Approved.

File: projects/krb5/build.sh [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+cp -r rdreq/ $OUT/.
+
+cp Fuzz_ndr $OUT/Fuzz_ndr
+cp Fuzz_pac $OUT/Fuzz_pac
+cp Fuzz_chpw $OUT/Fuzz_chpw
+cp Fuzz_json $OUT/Fuzz_json
+cp Fuzz_profile $OUT/Fuzz_profile
+cp Fuzz_marshal $OUT/Fuzz_marshal
+
+cp Fuzz_ndr_seed_corpus.zip $OUT/Fuzz_ndr_seed_corpus.zip
```
Comment (jonathanmetzman) [inline projects/krb5/build.sh:35]: Again, please download the zips from someplace else

**Review by jonathanmetzman — CHANGES_REQUESTED** [review]
lgtm other than the zips issue

**Review by jonathanmetzman — APPROVED** [review]
lgtm

**Review by jonathanmetzman — CHANGES_REQUESTED** [review]
Can you confirm upstream wants this?

**Comment by 0x34d** [comment]
> Hi! I no longer work on @krb5 for MIT, for over 5 years now. kaduk has been gone for even longer. Please explain how you decided which individuals to contact or list. Thanks.

@tlyu It's Simple, Are you an active member of the project right now?
And can you fix bugs produced by oss-fuzz in your project?

**Comment by greghudson** [comment]
Because of project staff turnover, we do not want an externally maintained list of individual maintainers.  Bugs should be reported to krbcore-security@mit.edu if they might be security-critical, or to krb5-bugs@mit.edu otherwise.

I'd also like to clarify an earlier statement about API surface.  There are a lot of API and protocol surfaces for krb5, only a few of which are touched on here.  Many of them are challenging for fuzzers because they use authenticated encryption, as any change to the ciphertext will result in a decryption failure.  So, for instance, the krb5_rd_req() harness will exercise some of the code, but everything that happens after decrypt_ticket() and decrypt_authenticator() will most likely only see the corpus's original data.  This problem is not unique to krb5, so perhaps oss-fuzz has a way of addressing it.

There are less severe challenges for surfaces using ASN.1 DER decoding.  Without specific knowledge of DER, a fuzzer which modifies a corpus of DER encodings will not produce a lot of valid encodings that change the original lengths of individual ASN.1 fields. as the DER tag-and-length prefixes on those fields won't match up.  This problem isn't unique either; it would manifest for any code base using length-prefixed encoding formats (such as protobufs).

**Comment by tlyu** [comment]
> > Hi! I no longer work on @krb5 for MIT, for over 5 years now. kaduk has been gone for even longer. Please explain how you decided which individuals to contact or list. Thanks.
> 
> @tlyu It's Simple, Are you an active member of the project right now? And can you fix bugs produced by oss-fuzz in your project?

Thank you for removing my address from the list. If those are your criteria, it is puzzling that you chose to list me.

I haven't been active in this project for years, which should be obvious from a casual inspection of the revision history.

Although I might theoretically have the skills and knowledge to fix bugs found in krb5 by oss-fuzz, I am not being funded to do so, and I also do not have current familiarity with the project's code or internals. I know that there might be some outdated web pages that refer to me, so I wonder if that's how you found me.

**Comment by 0x34d** [comment]
> I know that there might be some outdated web pages that refer to me, so I wonder if that's how you found me.

Yes exactly.

**Comment by 0x34d** [comment]
> Because of project staff turnover, we do not want an externally maintained list of individual maintainers. Bugs should be reported to [krbcore-security@mit.edu](mailto:krbcore-security@mit.edu) if they might be security-critical, or to [krb5-bugs@mit.edu](mailto:krb5-bugs@mit.edu) otherwise.

Ok, But to access the bug reports. These emails should have a GitHub account.


> I'd also like to clarify an earlier statement about API surface. There are a lot of API and protocol surfaces for krb5, only a few of which are touched on here. Many of them are challenging for fuzzers because they use authenticated encryption, as any change to the ciphertext will result in a decryption failure. So, for instance, the krb5_rd_req() harness will exercise some of the code, but everything that happens after decrypt_ticket() and decrypt_authenticator() will most likely only see the corpus's original data. This problem is not unique to krb5, so perhaps oss-fuzz has a way of addressing it.

I understand your concern about the Crypto part.
I'm not a Crypto person, learning and working with crypto protocol is on my list.
Possibly in the future, I will craft an [krb5] update PR regarding  Crypto functionality.
1. Fuzzing Crypto Function.
2. Fuzzing Crypto implementation.
3. Fuzzing encrypted data parser.

> There are less severe challenges for surfaces using ASN.1 DER decoding. Without specific knowledge of DER, a fuzzer which modifies a corpus of DER encodings will not produce a lot of valid encodings that change the original lengths of individual ASN.1 fields. as the DER tag-and-length prefixes on those fields won't match up. This problem isn't unique either; it would manifest for any code base using length-prefixed encoding formats (such as protobufs).

Got it, 
As I earlier said, I don't have a good hands-on experience in Cryptography.
If I even try surely I will make tons of mistakes in the implementation of a good guided encoding Crypto channel.

**Comment by 0x34d** [comment]
@jonathanmetzman 
I believe you can merge now.

**Comment by greghudson** [comment]
> Ok, But to access the bug reports. These emails should have a GitHub account.

Given the number of Github-averse projects in the OSS space, I am surprised to learn that oss-fuzz relies on Github accounts for bug communication.  We are not really interested in creating accounts for the bug-reporting contact addresses; we do (obviously) have the krb5 organization account if that helps.

**Comment by 0x34d** [comment]
> we do (obviously) have the krb5 organization account if that helps.

Alright, 
But you need to provide primary mail address of krb5 GitHub account,
for  Access Control List(ACL) in oss-fuzz.

**Comment by greghudson** [comment]
The corresponding email address is krbcore@mit.edu.  (We used to advertise this address as a point of contact for private communication with the krb5 team, but found that it was almost always misused, so it's not currently advertised as a place to send mail.)  I'm a little confused as to how github would use an email address rather than an account name  for access control, but I'll take your word for it.

**Pull request merged**